AI/Nullkiller/Analyzers/BuildAnalyzer.cpp

```diff
@@ -30,30 +31,22 @@ void BuildAnalyzer::updateTownDwellings(TownDevelopmentInfo & developmentInfo)
 	for(int level = 0; level < developmentInfo.town->getTown()->creatures.size(); level++)
 	{
 		logAi->trace("Checking dwelling level %d", level);
-		BuildingInfo nextToBuild = BuildingInfo();
-
-		BuildingID buildID = BuildingID(BuildingID::getDwellingFromLevel(level, 0));
-
-		for(; developmentInfo.town->getBuildings().count(buildID); BuildingID::advanceDwelling(buildID))
+		for(BuildingID buildID = BuildingID::getDwellingFromLevel(level, 0); buildID.hasValue(); BuildingID::advanceDwelling(buildID))
 		{
-			if(!developmentInfo.town->hasBuilt(buildID))
+			if(developmentInfo.town->getTown()->buildings.count(buildID) == 0)
```

Comment:
why does "hasBuilt" not suffice?

Reply:
Because "has built" is for "was this building constructed in town".
But this code actually needed to check for "does building of such type exists in this town"

Cove might have 2nd upgrade of 3rd dwelling built (or not built), but Castle does not have such dwelling to begin with.